cohortextractor/cohortextractor.py

```diff
@@ -772,6 +774,21 @@ def main():
         except DummyDataValidationError as e:
             print(f"Dummy data error: {e}")
             sys.exit(1)
+        except database_exceptions() as e:
+            traceback.print_exc()
+            # Exit with specific exit codes to help identify known issues
+            if "Unexpected EOF from the server" in str(e):
+                logger.error(f"Intermittent database error: {e}")
+                sys.exit(151)
```

Comment:
I don't think we should be using and exit code above 128, unless there's something I'm missing?

This is our process, we control the code, so we can decide what its exit codes mean.

So I would suggest we use 2, 3, 4, etc, for different error cases (maybe reserving exit code of 1 for unhandled errors).

Error codes above 128 typically mean that the process was killed externally by a signal, the exit code being 128 + signal number. So, and exit code 137 means the process was killed by signal number 9 (SIGKILL), of `kill -9` fame, which typically means the OOM killer, which users the most powerful signal available.

Reply:
yeah, max signal numbers and exit codes conventions vary a lot, so being conservative is a good idea, and that seems a reasonable range.

Running cohortextractor jobs in a BSD jail sounds like my weekend plans now :D

I've been assuming that the main goal of using different exit codes here is so that the job-runner executor can communicate something more useful to the user about the specifics of why their job failed. We could in theory have some convention of showing the last line of log output, or similar, which would mean any non-zero code would be fine. But then the old privacy leak issue rears its head, so I guess we need to communicate via ERRNOs like its 1989.

It would be really great if there was a way to safely communicate a short string from the docker process. I guess we could check for specific strings on the last few lines, which is almost as bad as semantic exit codes, but more flexible.

Reply:
I've updated this PR to look for the string "DatabaseError" instead of specific exceptions, and to use exit codes within the available range.  Couldn't quite figure out from the last couple of comments here if either of you were suggesting we should ditch this approach altogether.

Reply:
Thanks Becky, will take a look.

I was agreeing with @bloodearnest that this _is_ all a bit 1989, but also trying to argue that it's our best option at the moment. So unless Simon's got a brilliant leftfield idea up his sleeve (not a possibility to discount!) I think we should go ahead.

Reply:
Given that cohort-extractor's going away at some point, I think generally doing the easiest thing at the moment is probably best

Reply:
Sorry, was musing about an alternative, but @dave's point about executor API changes is valid, and I think exit codes is the only way for now, sorry for the confusion.